internal/pkg/agent/application/upgrade/marker_watcher_test.go

```diff
@@ -0,0 +1,84 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/details"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+func TestMarkerWatcher(t *testing.T) {
+	testMarkerDir := t.TempDir()
+	testMarkerFile := filepath.Join(testMarkerDir, markerFilename)
+	testLogger, _ := logger.NewTesting("watch_marker")
+
+	markerWatcher, err := newMarkerFileWatcher(testMarkerFile, testLogger)
+	require.NoError(t, err)
+	defer markerWatcher.Close()
+
+	testCtx, testCancel := context.WithCancel(context.Background())
+	defer testCancel()
+
+	var testDetails *details.Details
+	var testDetailsMu sync.Mutex
+
+	var testErr error
+	go func() {
+		for {
+			select {
+			case <-testCtx.Done():
+				return
+			case err := <-markerWatcher.Errors():
+				testErr = err
+			case marker := <-markerWatcher.Watch():
+				testDetailsMu.Lock()
+				testDetails = marker.Details
+				testDetailsMu.Unlock()
+			}
+		}
+	}()
+
+	go func() {
+		err := markerWatcher.Run(testCtx)
+		require.NoError(t, err)
```

Comment:
Because of `t.FailNow()` called by `require`, shouldn't the `require.*` be called from the goroutine running the test function?

Reply:
yeah, `require` does not work on a goroutine

Reply:
if you need to abort the test, you'd need to signal the test goroutine to exit

Reply:
Addressed as part of the refactoring done in 83577da728.